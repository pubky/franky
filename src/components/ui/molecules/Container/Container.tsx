import { ReactNode } from 'react';
import { cn } from '@/libs';

interface ContainerProps {
  as?: 'div' | 'section' | 'article' | 'main' | 'header' | 'footer' | 'aside' | 'nav' | 'body' | 'html' | 'figure';
  children?: ReactNode;
  className?: React.HTMLAttributes<HTMLDivElement>['className'];
  size?: 'sm' | 'md' | 'lg' | 'xl' | 'container' | 'default';
  display?: 'flex' | 'grid' | 'block';
}

export function Container({
  as: Tag = 'div',
  size = 'default',
  display = 'flex',
  ...props
}: ContainerProps &
  React.HTMLAttributes<HTMLDivElement> &
  React.HTMLAttributes<HTMLHtmlElement> &
  React.HTMLAttributes<HTMLBodyElement>) {
  const defaultClasses = 'mx-auto w-full flex-col';
  const displayClasses = {
    flex: 'flex',
    grid: 'grid',
    block: 'block',
  };
  const sizeClasses = {
    default: '',
    sm: 'max-w-screen-sm',
    md: 'max-w-screen-md',
    lg: 'max-w-screen-lg',
    xl: 'max-w-screen-xl',
    container: 'container',
  };

  const containerClassName = cn(defaultClasses, displayClasses[display], sizeClasses[size], props.className);

  return (
    <Tag {...props} className={containerClassName}>
      {props.children}
    </Tag>
  );
}
