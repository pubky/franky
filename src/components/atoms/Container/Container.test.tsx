import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Container } from './Container';

describe('Container', () => {
  it('renders with default props', () => {
    render(<Container data-testid="container">Test content</Container>);

    const container = screen.getByTestId('container');
    expect(container).toBeInTheDocument();
    expect(container.tagName).toBe('DIV');
    expect(container).toHaveClass('mx-auto', 'w-full', 'flex-col', 'flex');
    expect(screen.getByText('Test content')).toBeInTheDocument();
  });

  it('renders different HTML elements', () => {
    const { rerender } = render(
      <Container as="section" data-testid="container">
        Content
      </Container>,
    );

    let container = screen.getByTestId('container');
    expect(container.tagName).toBe('SECTION');

    rerender(
      <Container as="main" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container.tagName).toBe('MAIN');
  });

  it('renders different sizes correctly', () => {
    const { rerender } = render(
      <Container size="sm" data-testid="container">
        Content
      </Container>,
    );

    let container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-screen-sm');

    rerender(
      <Container size="md" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-screen-md');

    rerender(
      <Container size="lg" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-screen-lg');

    rerender(
      <Container size="xl" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-screen-xl');

    rerender(
      <Container size="container" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('container');
  });

  it('renders different display modes correctly', () => {
    const { rerender } = render(
      <Container display="grid" data-testid="container">
        Content
      </Container>,
    );

    let container = screen.getByTestId('container');
    expect(container).toHaveClass('grid');

    rerender(
      <Container display="block" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('block');

    rerender(
      <Container display="flex" data-testid="container">
        Content
      </Container>,
    );
    container = screen.getByTestId('container');
    expect(container).toHaveClass('flex');
  });

  it('applies custom className', () => {
    render(
      <Container className="custom-container" data-testid="container">
        Content
      </Container>,
    );

    const container = screen.getByTestId('container');
    expect(container).toHaveClass('custom-container');
  });

  it('combines all props correctly', () => {
    render(
      <Container as="article" size="lg" display="grid" className="custom" data-testid="container">
        Combined props content
      </Container>,
    );

    const container = screen.getByTestId('container');
    expect(container.tagName).toBe('ARTICLE');
    expect(container).toHaveClass('mx-auto', 'w-full', 'flex-col', 'grid', 'max-w-screen-lg', 'custom');
  });
});
