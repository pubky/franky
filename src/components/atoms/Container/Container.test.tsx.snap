// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Container - Snapshots > matches snapshot for article element 1`] = `
<article
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Article Container
</article>
`;

exports[`Container - Snapshots > matches snapshot for block display 1`] = `
<div
  class="mx-auto w-full flex-col block"
  data-testid="container"
>
  Block Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for container size 1`] = `
<div
  class="mx-auto w-full flex-col flex container"
  data-testid="container"
>
  Container Size
</div>
`;

exports[`Container - Snapshots > matches snapshot for default size 1`] = `
<div
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Default Size
</div>
`;

exports[`Container - Snapshots > matches snapshot for div element 1`] = `
<div
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Div Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for extra large size 1`] = `
<div
  class="mx-auto w-full flex-col flex max-w-screen-xl"
  data-testid="container"
>
  Extra Large Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for flex display 1`] = `
<div
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Flex Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for footer element 1`] = `
<footer
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Footer Container
</footer>
`;

exports[`Container - Snapshots > matches snapshot for grid display 1`] = `
<div
  class="mx-auto w-full flex-col grid"
  data-testid="container"
>
  Grid Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for header element 1`] = `
<header
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Header Container
</header>
`;

exports[`Container - Snapshots > matches snapshot for large size 1`] = `
<div
  class="mx-auto w-full flex-col flex max-w-screen-lg"
  data-testid="container"
>
  Large Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for main element 1`] = `
<main
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Main Container
</main>
`;

exports[`Container - Snapshots > matches snapshot for medium size 1`] = `
<div
  class="mx-auto w-full flex-col flex max-w-screen-md"
  data-testid="container"
>
  Medium Container
</div>
`;

exports[`Container - Snapshots > matches snapshot for section element 1`] = `
<section
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Section Container
</section>
`;

exports[`Container - Snapshots > matches snapshot for small size 1`] = `
<div
  class="mx-auto w-full flex-col flex max-w-screen-sm"
  data-testid="container"
>
  Small Container
</div>
`;

exports[`Container - Snapshots > matches snapshot with default props 1`] = `
<div
  class="mx-auto w-full flex-col flex"
  data-testid="container"
>
  Default Container
</div>
`;

exports[`Container - Snapshots > matches snapshots for combined props 1`] = `
<article
  class="mx-auto w-full flex-col grid max-w-screen-lg custom-class"
  data-testid="container"
>
  Combined Props Container
</article>
`;
